crypto/asn1/tasn_dec.c

```diff
@@ -855,19 +911,89 @@ static int asn1_ex_c2i(ASN1_VALUE **pval, const unsigned char *cont, long len,
   return ret;
 }
 
-// Check an ASN1 tag and length: a bit like ASN1_get_object but it
-// checks the expected tag.
+// This function finds the end of an ASN1 structure when passed its maximum
+// length, whether it is indefinite length and a pointer to the content. This
+// is more efficient than calling asn1_collect because it does not recurse on
+// each indefinite length header.
+static int asn1_find_end(const unsigned char **in, long len, char inf) {
+  uint32_t expected_eoc;
+  long plen;
+  const unsigned char *p = *in, *q;
+  // If not indefinite length constructed just add length.
+  if (inf == 0) {
+    *in += len;
+    return 1;
+  }
+  expected_eoc = 1;
+  // Indefinite length constructed form. Find the end when enough EOCs are
+  // found. If more indefinite length constructed headers are encountered
+  // increment the expected eoc count otherwise just skip to the end of the
+  // data.
+  while (len > 0) {
+    if (asn1_check_eoc(&p, len)) {
+      expected_eoc--;
+      if (expected_eoc == 0) {
+        break;
+      }
+      len -= 2;
+      continue;
+    }
+    q = p;
+    // Just read in a header: only care about the length.
+    if (!asn1_check_tlen(&plen, NULL, NULL, &inf, NULL, &p, len,
+                         -1, 0, 0)) {
+      OPENSSL_PUT_ERROR(ASN1, ASN1_R_NESTED_ASN1_ERROR);
+      return 0;
+    }
+    if (inf) {
+      if (expected_eoc == UINT32_MAX) {
```

Comment:
nit: suggest adding a comment explaining that this is an underflow, and the significance of that condition in the loop